src/OrchardCore/OrchardCore/DefaultKeyedServiceResolver.cs

```diff
@@ -0,0 +1,16 @@
+using System;
+using System.Collections.Frozen;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace OrchardCore;
+
+public class DefaultKeyedServiceResolver(IServiceCollection services) : IKeyedServiceResolver
+{
+    private readonly IServiceCollection _services = services;
+
+    public IReadOnlyDictionary<TKey, TService> GetServices<TKey, TService>()
+        => _services.Where(service => service.ServiceKey is not null && service.ServiceKey!.GetType() == typeof(TKey) && service.ServiceType == typeof(TService))
+                 .ToFrozenDictionary(service => (TKey)Convert.ChangeType(service.ServiceKey, typeof(TKey)), service => (TService)service.KeyedImplementationInstance);
+}
```

Comment:
Yes, for example when creating a child container the following returns an IEnumerable.

    serviceProvider.GetKeyedServices(services.Key.ServiceType, services.Key.ServiceKey)

Reply:
@jtkech Should we return an array of document that exposes the key and the instance instead?

Reply:
What does not make since to me is that when you resolve an instance by key, you always get a single instance. For example resolving using FromKeyedServicesAttribute, you get 1 instance for that key. Not a collection. I think we should be fine using Dictionary

Reply:
@jtkech I made some changes here. 
- The `GetServices<TKey, TService>()` now returns `IEnumerable<KeyValuePair<TKey, TService>>` so we can handle the scenario where one key would return multiple instances.
- Also, keyed services are now resolved using the `IServiceProvider` instead of trying to get the instances out of `IServiceCollection`.
- Added `GetServicesAsDictionary<TKey, TService>` extension that would resolved `KeyedServiceDictionary<TKey, TService>` using the last instance as done in `FromKeyedServicesAttribute`
- I register the `IServiceCollection` into the IoC so that we can access it later to obtain into about available keys.


Let me know your thoughts about the updated approach.